internal/backend/runtime/omni/controllers/omni/discovery_audit.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic/qtransform"
+	"github.com/cosi-project/runtime/pkg/resource"
+	serverpb "github.com/siderolabs/discovery-api/api/v1alpha1/server/pb"
+	"github.com/siderolabs/gen/xerrors"
+	"github.com/siderolabs/gen/xslices"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+)
+
+// DiscoveryAuditController creates the system config patch that contains the maintenance config.
+type DiscoveryAuditController = qtransform.QController[*omni.ClusterIdentity, *omni.DiscoveryAuditResult]
+
+// NewDiscoveryAuditController initializes DiscoveryAuditController.
+func NewDiscoveryAuditController(deleteOlderThan time.Duration) *DiscoveryAuditController {
+	requeueAfter := deleteOlderThan + 1*time.Second
+	auditor := &discoveryAuditor{
+		deleteOlderThan: deleteOlderThan,
+	}
+
+	return qtransform.NewQController(
+		qtransform.Settings[*omni.ClusterIdentity, *omni.DiscoveryAuditResult]{
+			Name: "DiscoveryAuditController",
+			MapMetadataFunc: func(clusterIdentity *omni.ClusterIdentity) *omni.DiscoveryAuditResult {
+				return omni.NewDiscoveryAuditResult(resources.DefaultNamespace, clusterIdentity.Metadata().ID())
+			},
+			UnmapMetadataFunc: func(discoveryAuditResult *omni.DiscoveryAuditResult) *omni.ClusterIdentity {
+				return omni.NewClusterIdentity(resources.DefaultNamespace, discoveryAuditResult.Metadata().ID())
+			},
+			TransformFunc: func(ctx context.Context, _ controller.Reader, logger *zap.Logger, clusterIdentity *omni.ClusterIdentity, discoveryAuditResult *omni.DiscoveryAuditResult) error {
+				discoveryClientConn, err := createDiscoveryClientConn(ctx)
+				if err != nil {
+					return fmt.Errorf("failed to create discovery client: %w", err)
+				}
+
+				defer func() {
+					if closeErr := discoveryClientConn.Close(); closeErr != nil {
+						logger.Error("failed to close discovery client", zap.Error(closeErr))
+					}
+				}()
+
+				discoveryClient := serverpb.NewClusterClient(discoveryClientConn)
+				clusterID := clusterIdentity.TypedSpec().Value.ClusterId
+
+				logger = logger.With(zap.String("cluster_id", clusterID))
+
+				discoveryAffiliateIDs, err := auditor.listAffiliateIDs(ctx, discoveryClient, clusterID)
+				if err != nil {
+					return err
+				}
+
+				var localNodeIDs []string
+
+				if clusterIdentity.Metadata().Phase() != resource.PhaseTearingDown { // if the cluster is being torn down, we consider all nodes to be invalid
```

Comment:
can we do potentially something more simple, i.e. drop affiliates for ClusterIdentities which are being torn down?

Reply:
Done, and reworked the whole thing quite a bit. This was wrong anyway - it needed to be in `FinalizerRemoval`.

(still a draft, will add a couple of more things before it's ready for review).